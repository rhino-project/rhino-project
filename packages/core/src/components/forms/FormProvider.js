import PropTypes from 'prop-types';
import { FormProvider as RHFFormProvider } from 'react-hook-form';
import { DevTool } from '@hookform/devtools';

const FormProvider = ({ children, ...props }) => {
  const { control } = props;

  return (
    <>
      {import.meta.env.DEV && (
        <DevTool
          control={control}
          placement={'top-left'}
          styles={{
            button: {
              width: '20px'
            }
          }}
        />
      )}
      <RHFFormProvider {...props}>{children}</RHFFormProvider>
    </>
  );
};

FormProvider.propTypes = {
  children: PropTypes.node
};

export default FormProvider;
