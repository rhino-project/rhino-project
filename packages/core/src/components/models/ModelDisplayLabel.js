import { useMemo } from 'react';
import { useGlobalComponent } from 'rhino/hooks/overrides';
import DisplayLabel from '../forms/DisplayLabel';
import { useModelAndAttributeFromPath } from 'rhino/hooks/models';

export const ModelDisplayLabelBase = ({
  overrides,
  label,
  model,
  ...props
}) => {
  const { attribute } = useModelAndAttributeFromPath(model, props.path);

  const modelLabel = useMemo(() => label || attribute.readableName, [
    attribute,
    label
  ]);

  return <DisplayLabel label={modelLabel} {...props} />;
};

const ModelDisplayLabel = (props) =>
  useGlobalComponent('ModelDisplayLabel', ModelDisplayLabelBase, props);

export default ModelDisplayLabel;
