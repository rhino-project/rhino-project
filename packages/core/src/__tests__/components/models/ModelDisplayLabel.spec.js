import { render } from '@testing-library/react';
import { useForm } from 'react-hook-form';
import FormProvider from 'rhino/components/forms/FormProvider';
import ModelDisplayLabel from 'rhino/components/models/ModelDisplayLabel';
import { sharedModelTests } from './sharedModelTests';

describe('ModelDisplayLabel', () => {
  const wrapper = ({ children }) => {
    const methods = useForm();
    return <FormProvider {...methods}>{children}</FormProvider>;
  };

  sharedModelTests(ModelDisplayLabel);

  it(`should render label with readable name`, () => {
    const { asFragment } = render(
      <ModelDisplayLabel
        model={{
          properties: { dummy: { type: 'string', readableName: 'Dummy' } }
        }}
        path="dummy"
      />,
      {
        wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it(`should override label with prop`, () => {
    const { asFragment } = render(
      <ModelDisplayLabel
        model={{
          properties: { dummy: { type: 'string', readableName: 'Dummy' } }
        }}
        label="Dummy Override"
        path="dummy"
      />,
      {
        wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
