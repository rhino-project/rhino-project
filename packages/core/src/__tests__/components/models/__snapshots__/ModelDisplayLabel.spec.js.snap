// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelDisplayLabel > should override label with prop 1`] = `
<DocumentFragment>
  <label
    class="form-label"
    for="dummy"
    id="displayLabel-dummy"
    path="dummy"
  >
    Dummy Override
  </label>
</DocumentFragment>
`;

exports[`ModelDisplayLabel > should render label with readable name 1`] = `
<DocumentFragment>
  <label
    class="form-label"
    for="dummy"
    id="displayLabel-dummy"
    path="dummy"
  >
    Dummy
  </label>
</DocumentFragment>
`;

exports[`ModelDisplayLabel > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelDisplayLabel > should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
